model/user/user.go

```diff
@@ -208,6 +209,26 @@ func (u *DBUser) IncPatchNumber() (int, error) {
 	return dbUser.PatchNumber, nil
 }
 
+func (u *DBUser) AddFavoritedProject(identifier string) ([]string, error) {
+	if util.StringSliceContains(u.FavoriteProjects, identifier) {
+		return nil, errors.Errorf("cannot add duplicate project '%s'", identifier)
+	}
+	update := bson.M{
+		"$push": bson.M{FavoriteProjectsKey: identifier},
+	}
+	if err := UpdateOne(bson.M{IdKey: u.Id}, update); err != nil {
+		return nil, err
+	}
+	u.FavoriteProjects = append(u.FavoriteProjects, identifier)
+
+	err := event.LogUserEvent(u.Id, event.UserEventTypeFavoriteProjectsUpdate, u.FavoriteProjects[:len(u.FavoriteProjects)-1], u.FavoriteProjects)
```

Comment:
I would save the before state in a variable before it's changed, rather than slicing the array afterwards

Reply:
for some reason it's done this way in all the other funcs in this module which is where I got it from. will change